### ISSUE TITLE
Expected translation model training speed-up with apex

### ISSUE THREAD

**Issue by stephanpeitz** [pytorch/fairseq#3174]
#### What is your question?

Hi all,

what is the expected training speed-up once apex is activated?
I installed the apex library as recommend, however, I don't see any speed-up for my setup.

#### What have you tried?

I compared two setups, one with apex library installed, the other without.
The command looks like: 
` fairseq-train data-bin --log-format simple --log-interval 100 --save-dir /output --max-tokens 7168 --update-freq 1                                          --num-workers 1 --arch transformer --optimizer adam --adam-betas '(0.9,0.98)' --clip-norm 0.0 --lr-scheduler inverse_sqrt --warmup-init-lr 1e-07 --warmup-updates 4000 --lr  0.0007 --min-lr 1e-09 --dropout 0.1 --attention-dropout 0.0 --weight-decay 0.0 --criterion label_smoothed_cross_entropy --label-smoothing 0.0 --fp16 --seed 2 --max-epoch 0 --task translation  --ddp-backend=no_c10d --max-update 80000 --save-interval-updates 2000 --no-epoch-checkpoints --share-all-embeddings --criterion label_smoothed_cross_entropy -distributed-world-size 32 --distributed-backend 'nccl'  --distributed-init-method 'tcp://XXX' --distributed-no-spawn`

Successful installation is confirmed with:
`Running setup.py install for apex: finished with status 'done'`
`(self_attn_layer_norm): FusedLayerNorm(torch.Size([512]), eps=1e-05, elementwise_affine=True)`
`2021-01-27 15:52:48 | INFO | fairseq.optim.adam | using FusedAdam`

#### What's your environment?

 - fairseq Version: 0.10.1
 - PyTorch Version: 1.8.0.dev20210127 (with conda)
 - OS: Linux
 - How you installed fairseq: `pip install --editable .`
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1.243/7.6.3
 - GPU models and configuration: 32x V100 (4x 8 GPU nodes)
 - Any other relevant information: cudatoolkit(-dev for nvcc) was installed with conda

What do I miss? 
Any experience with apex and a medium-sized transformer-based NMT setup?

Thanks for your help!

Cheers,
Stephan

**Comment by lematt1991** [comment]
CC @myleott

**Comment by myleott** [comment]
It should be a few percent faster. Here's what I get on a single machine with 8xV100:

**command**
```
python -m fairseq_cli.train --task dummy_mt --dataset-size 10000000 \
  --arch transformer_wmt_en_de_big --max-tokens 4096 \
  --optimizer adam --lr 0.0001 \
  --log-format simple --log-interval 25 \
  --no-save --max-update 500 \
  --fp16
```

**without apex**:
```
2021-01-29 05:49:31 | INFO | fairseq.trainer | NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 64.0
2021-01-29 05:49:33 | INFO | train_inner | epoch 001:     26 / 9192 loss=4.831, ppl=28.46, wps=171455, ups=5.25, wpb=32640, bsz=1088, num_updates=25, lr=0.0001, gnorm=8.71, loss_scale=64, train_wall=6, gb_free=20.9, mb_free=21445.4, wall=28
2021-01-29 05:49:37 | INFO | train_inner | epoch 001:     51 / 9192 loss=0.006, ppl=1, wps=186646, ups=5.72, wpb=32640, bsz=1088, num_updates=50, lr=0.0001, gnorm=0.105, loss_scale=64, train_wall=4, gb_free=20.9, mb_free=21445.4, wall=33
2021-01-29 05:49:42 | INFO | train_inner | epoch 001:     76 / 9192 loss=0, ppl=1, wps=186931, ups=5.73, wpb=32640, bsz=1088, num_updates=75, lr=0.0001, gnorm=0.007, loss_scale=64, train_wall=4, gb_free=20.9, mb_free=21445.4, wall=37
2021-01-29 05:49:46 | INFO | train_inner | epoch 001:    101 / 9192 loss=0, ppl=1, wps=186816, ups=5.72, wpb=32640, bsz=1088, num_updates=100, lr=0.0001, gnorm=0.003, loss_scale=64, train_wall=4, gb_free=20.9, mb_free=21445.4, wall=41
```

**with apex**:
```
2021-01-29 05:55:25 | INFO | fairseq.trainer | NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 64.0
2021-01-29 05:55:27 | INFO | train_inner | epoch 001:     26 / 9192 loss=4.831, ppl=28.46, wps=179591, ups=5.5, wpb=32640, bsz=1088, num_updates=25, lr=0.0001, gnorm=8.833, loss_scale=64, train_wall=6, gb_free=20.9, mb_free=21444.9, wall=28
2021-01-29 05:55:31 | INFO | train_inner | epoch 001:     51 / 9192 loss=0.006, ppl=1, wps=193285, ups=5.92, wpb=32640, bsz=1088, num_updates=50, lr=0.0001, gnorm=0.109, loss_scale=64, train_wall=4, gb_free=20.9, mb_free=21444.9, wall=32
2021-01-29 05:55:36 | INFO | train_inner | epoch 001:     76 / 9192 loss=0, ppl=1, wps=193434, ups=5.93, wpb=32640, bsz=1088, num_updates=75, lr=0.0001, gnorm=0.007, loss_scale=64, train_wall=4, gb_free=20.9, mb_free=21444.9, wall=36
2021-01-29 05:55:40 | INFO | train_inner | epoch 001:    101 / 9192 loss=0, ppl=1, wps=184334, ups=5.65, wpb=32640, bsz=1088, num_updates=100, lr=0.0001, gnorm=0.003, loss_scale=64, train_wall=4, gb_free=20.9, mb_free=21444.9, wall=41
2021-01-29 05:55:44 | INFO | train_inner | epoch 001:    126 / 9192 loss=0, ppl=1, wps=193405, ups=5.93, wpb=32640, bsz=1088, num_updates=125, lr=0.0001, gnorm=0.002, loss_scale=64, train_wall=4, gb_free=20.9, mb_free=21444.9, wall=45
2021-01-29 05:55:48 | INFO | train_inner | epoch 001:    151 / 9192 loss=0, ppl=1, wps=193741, ups=5.94, wpb=32640, bsz=1088, num_updates=150, lr=0.0001, gnorm=0.002, loss_scale=64, train_wall=4, gb_free=20.9, mb_free=21444.9, wall=49
```

**Comment by stephanpeitz** [comment]
Hey,

thanks for the quick and detailed answer!
Looks like 3-4% in your case. I observe 1-2%.
Maybe the speed-up is smaller since I train already on 32 GPUs.

Cheers,
Stephan
